sql/core/src/main/scala/org/apache/spark/sql/execution/basicOperators.scala

```diff
@@ -36,11 +36,15 @@ import org.apache.spark.{HashPartitioner, SparkEnv}
 case class Project(projectList: Seq[NamedExpression], child: SparkPlan) extends UnaryNode {
   override def output: Seq[Attribute] = projectList.map(_.toAttribute)
 
-  @transient lazy val buildProjection = newMutableProjection(projectList, child.output)
+  private def buildProjection = newMutableProjection(projectList, child.output)
 
-  protected override def doExecute(): RDD[InternalRow] = child.execute().mapPartitions { iter =>
-    val reusableProjection = buildProjection()
-    iter.map(reusableProjection)
+  protected override def doExecute(): RDD[InternalRow] = {
+    // Use local variable to avoid referencing to $out inside closure
```

Comment:
Though is this even a hot code path? The only reason why we do it in `DataSourceStrategy` is because we call `mapPartitions` once for each data partition there is (could be O(1000)). IIUC this is only called once or twice per query. Is that correct @yhuai? If so I'm not sure if the optimization from `DataSourceStrategy` is worth doing here.